engine/gamesys/src/gamesys/scripts/script_particlefx.cpp

```diff
@@ -75,52 +75,45 @@ namespace dmGameSystem
      * @variable
      */
 
-    struct EmitterStateChangedCallbackArgs
-    {
-        EmitterStateChangedCallbackArgs(dmhash_t component_id, dmhash_t emitter_id, dmParticle::EmitterState emitter_state)
-            : m_ComponentId(component_id), m_EmitterId(emitter_id), m_EmitterState(emitter_state)
-        {}
-        dmhash_t m_ComponentId;
-        dmhash_t m_EmitterId;
-        dmParticle::EmitterState m_EmitterState;
-    };
-
-    static void EmitterStateChangedCallbackArgsCB(lua_State* L, void* user_args)
-    {
-        EmitterStateChangedCallbackArgs* args = (EmitterStateChangedCallbackArgs*)user_args;
-        dmScript::PushHash(L, args->m_ComponentId);
-        dmScript::PushHash(L, args->m_EmitterId);
-        lua_pushnumber(L, args->m_EmitterState);
-    }
-
     void EmitterStateChangedCallback(uint32_t num_awake_emitters, dmhash_t emitter_id, dmParticle::EmitterState emitter_state, void* user_data)
     {
+
         EmitterStateChangedScriptData data = *(EmitterStateChangedScriptData*)(user_data);
 
-        if (data.m_CallbackInfo)
+        if (!data.m_CallbackInfo)
         {
-            if (!dmScript::IsCallbackValid(data.m_CallbackInfo))
-            {
-                return;
-            }
+            dmLogError("No callback set for particlefx.");
+            return;
+        }
+
+        if (!dmScript::IsCallbackValid(data.m_CallbackInfo))
+        {
+            return;
+        }
 
-            EmitterStateChangedCallbackArgs args(data.m_ComponentId, emitter_id, emitter_state);
+        lua_State* L = dmScript::GetCallbackLuaContext(data.m_CallbackInfo);
 
-            if (!dmScript::InvokeCallback(data.m_CallbackInfo, EmitterStateChangedCallbackArgsCB, &args))
-            {
-                dmLogError("Could not run particlefx callback because the instance has been deleted.");
-            }
+        DM_LUA_STACK_CHECK(L, 0);
 
-            // The last emitter belonging to this particlefx har gone to sleep, release lua reference.
-            if(num_awake_emitters == 0 && emitter_state == dmParticle::EMITTER_STATE_SLEEPING)
-            {
-                dmScript::DestroyCallback(data.m_CallbackInfo);
-                data.m_CallbackInfo = 0x0;
-            }
+        if (!dmScript::SetupCallback(data.m_CallbackInfo))
+        {
+            dmScript::DestroyCallback(data.m_CallbackInfo);
+            data.m_CallbackInfo = 0x0;
+            return;
         }
-        else
+
+        dmScript::PushHash(L, data.m_ComponentId);
+        dmScript::PushHash(L, emitter_id);
+        lua_pushnumber(L, emitter_state);
+        dmScript::PCall(L, 4, 0);
+
```

Comment:
Not sure if there's a point in checking for an error here? Since it's a Lua error in the actual callback function it might actually work the next time so I don't think we should destroy the callback.